File: elixir/boutique-inventory/lib/boutique_inventory.ex
defmodule BoutiqueInventory do
  def sort_by_price(inventory) do
    Enum.sort_by(inventory, & &1.price, :asc)
  end

  def with_missing_price(inventory) do
    Enum.filter(inventory, &is_nil(&1.price))
  end

  def update_names(inventory, search, replace) do
    Enum.map(
      inventory,
      fn item ->
        %{item | name: String.replace(item.name, search, replace)}
      end
    )
  end

  def increase_quantity(item, count) do
    update_in(
      item,
      [:quantity_by_size],
      fn qs ->
        Map.new(
          qs,
          fn {size, num} -> {size, num + count} end
        )
      end
    )
  end

  def total_quantity(item) do
    Enum.reduce(
      item.quantity_by_size,
      0,
      fn {_, num}, total -> total + num end
    )
  end
end

File: ruby/boutique-inventory/boutique_inventory.rb
class BoutiqueInventory
  PRICE_CHEAP_UPPER_LIMIT = 30
  
  def initialize(items)
    @items = items
  end

  def item_names
    items.map {|item| item[:name]}.sort!
  end

  def cheap
    items.select {|item| item[:price] < PRICE_CHEAP_UPPER_LIMIT}
  end

  def out_of_stock
    items.select {|item| item[:quantity_by_size].empty?}
  end

  def stock_for_item(name)
    (items.select {|item| item[:name] == name}.first || {})[:quantity_by_size]
  end

  def total_stock
    items.reduce(0) {|total, item| total + item[:quantity_by_size].values.sum}
  end

  private
  attr_reader :items
 end
